### DISCUSSION TITLE
Listing undefined symbols of a MathNode

### DISCUSSION THREAD

**Discussion by zopieux** [josdejong/mathjs#2551] [Q&A] [answered]
Hi there,

I'm using mathjs to pre-compile (parse) a bunch of expressions. I'm left with a list of `MathNode`s. I need to get for each of them, post-`simplify`, what "free-standing" symbols it references aka depends on. The attempt helper below:

    function symbols(node: MathNode): Set<string> {
      const symbols = new Set<string>()
      node.traverse((n) => (n.type == 'SymbolNode') && symbols.add(n.name))
      return symbols
    }

gives both the free-standing symbols *but also the builtin maths functions*:

    symbols(math.simplify('a+b+sqrt(b)+asin(c)'))
    // a, b, sqrt, asin, c
    // wanted instead: a, b, c

I could not find such an API in mathjs, does it exist? [Here is equivalent nerdamer function that I need](https://github.com/jiggzson/nerdamer/blob/master/nerdamer.core.js#L368=). Otherwise, as workaround, what would be a reliable way to exclude all builtins from the symbol list (ie. where are they stored please)?

Thanks!

**Comment by josdejong** [comment, marked as answer]
In the parsed expression there is indeed no difference between functions and symbols, nor between functions/symbols defined in mathjs or in your `scope`. Some tips:

- You can recognize function calls like `sqrt` and `asin` from symbols because they are defined as a `FunctionNode` having a property `.fn` being a `SymbolNode` with a `.name` property like `sqrt` or `asin`. You can use the third `parent` argument of `node.traverse(function (node, path, parent) {...})` to check that.
- Whether something is a built-in function of mathjs can be verified by checking whether say `math.asin` and `math.pi` are defined.
- It is possible that a symbol is a unit like `inch` or `millimeter`, which will be resolved by mathjs when evaluating. To verify that, you can check the symbol against `math.Unit.isValuelessUnit(name)`, which returns true for valid unit names.
- Not sure if relevant to you, but there are two static functions `FunctionNode.onUndefinedFunction(name)` and `SymbolNode.onUndefinedSymbol(name)` which can be overwritten to customize what to do when an undefined symbol is encountered whilst evaluating. I see these functions are not documented, I'll add them to the docs.

**Reply by zopieux** [reply]
Thanks for the reply and for improving the docs!
